### ISSUE TITLE
LLVM invalid memory load error

### ISSUE THREAD

**Issue by lcasburn** [GaloisInc/crucible#32]
The program below causes the LLVM symbolic simulator to fail to load a value from memory. When the string `s` is set to `foobar`, it prints the bytes of the string and then exits. 

```c
#include <stdio.h>
int main() {

  unsigned char s[] = "foobarb"; // fails
  //unsigned char s[] = "foobar";  // works

  for (int i=0; s[i] !=0; ++i)
    printf("byte %d = %#x\n", i, s[i]);

  printf("Moving printf before loop changes index of error: '%s'\n", s);
  return 0;
}
```

When `s` is set to `foobarb`, the first byte of the string is printed and then an invalid memory error is generated.
```
byte 0 = 0x66
Invalid memory load: address (7, 0x8:[64], 0x1:[64]) at type i8
in main at ./read-bytes.c:7:17
  When calling llvm_load
```

The LLVM bitcode in the successful case identifies `s` as a pointer. In the failing case, it stores the string as a 64bit integer and bitcasts.  One idea @robdockins wanted to track down was to see if this was related to read failing to find the overlapping memory region it belonged to.

Further complicating this example, is that if you move the `printf` on line 10 to the line before the loop, the invalid memory load location changes to:
```
Invalid memory load: address (7, 0x8:[64], 0x0:[64]) at type i8
in main at ./read-bytes.c:7:3
  When calling printf
```

Commenting out the line that prints the whole string generates a translation error indicating that the arguments to `llvm.dbg.declare` are ill-formed. For these examples, I used clang with llvm-3.6.2 and compiled with `-O3` optimizations.

**Comment by robdockins** [comment]
I can reproduce this error using clang 3.8.1.  Interestingly, the problem doesn't occur with `-O0`, and slightly different wrong behaviors occur with `-O1` as opposed to `-O2` and `-O3`.

**Comment by robdockins** [comment]
I've located the source of the bug.  It's a rather embarrassing situation of confusing "number of bytes" with "number of bits".  In some code that implements the necessary operations to select a subrange of bytes from a larger bitvector, some checks are necessary to ensure that the subrange selection is valid.  These checks were failing because of the unit confusion and falling through to a failure case.

Fixes are forthcoming.

**Comment by robdockins** [comment]
Fixed in 54e05eb6230c2a6383badb6c04797fd19523ca2d
